src/Services/SearchService.php

```diff
@@ -249,4 +260,43 @@ private function filterInvalidVariationIds(array $ids)
             return $result['total'] > 0;
         }));
     }
+
+    private function isRedirectToProductPage(array $productsIds): bool
+    {
+        if (count($productsIds) !== 1) {
+            return false;
+        }
+
+        $dataQueryInfoMessage = $this->getResults()->getData(Response::DATA_QUERY_INFO_MESSAGE);
+
+        $type = !empty($dataQueryInfoMessage['didYouMeanQuery'])
+            ? 'did-you-mean' : $dataQueryInfoMessage['queryStringType'];
+
+        return $type !== 'corrected' && $type !== 'improved';
+    }
+
+    /**
+     * @return string|null
+     */
+    private function getProductRedirectUrl(int $productId)
```

Comment:
Add the return type. E.g. `?string`.

Reply:
This actually gives you the product detail page url, not a redirect URL, because you could theoretically use it for any other purpose than redirecting. Rename this to `getProductDetailUrl`.